lib/segment/src/index/hnsw_index/graph_links.rs

```diff
@@ -328,55 +405,49 @@ pub struct GraphLinksRam {
     // all ranges in `links`. each range is `links[offsets[i]..offsets[i+1]]`
     // ranges are sorted by level
     offsets: Vec<u64>,
-    // start offet of each level in `offsets`
+    // start offset of each level in `offsets`
     level_offsets: Vec<u64>,
     // for level 1 and above: reindex[point_id] = index of point_id in offsets
     reindex: Vec<PointOffsetType>,
 }
 
+impl GraphLinksRam {
+    pub fn load_from_memory(data: &[u8]) -> Self {
+        let header = GraphLinksFileHeader::deserialize_bytes_from(data);
+        let links = links_slice(data, &header).to_vec();
+        let offsets = offsets_slice(data, &header).to_vec();
+        let level_offsets = level_offsets(data, &header);
+        let reindex = reindex_slice(data, &header).to_vec();
+        Self {
+            links,
+            offsets,
+            level_offsets,
+            reindex,
+        }
+    }
+}
+
 impl GraphLinks for GraphLinksRam {
     fn load_from_file(path: &Path) -> OperationResult<Self> {
-        let mmap = GraphLinksMmap::load_from_file(path)?;
-
-        Ok(Self {
-            links: mmap.get_links_slice().to_vec(),
-            offsets: mmap.get_offsets_slice().to_vec(),
-            level_offsets: mmap.level_offsets.clone(),
-            reindex: mmap.get_reindex_slice().to_vec(),
-        })
-    }
+        let file = OpenOptions::new()
+            .read(true)
+            .write(false)
+            .create(false)
+            .open(path)?;
 
-    fn from_vec(
-        edges: Vec<Vec<Vec<PointOffsetType>>>,
-        path: Option<&Path>,
-    ) -> OperationResult<Self> {
-        let mut graph_links = GraphLinksRam {
-            links: Vec::new(),
-            offsets: vec![0],
-            level_offsets: Vec::new(),
-            reindex: Vec::new(),
-        };
+        let mmap = unsafe { Mmap::map(&file)? };
 
-        let mut converter = GraphLinksConverter::new(edges);
-        if let Some(path) = path {
-            converter.save_to_file(path)?;
-        }
+        Ok(Self::load_from_memory(&mmap))
+    }
 
-        if !converter.edges.is_empty() {
-            let levels_count = converter.get_levels_count();
-            for level in 0..levels_count {
-                graph_links
-                    .level_offsets
-                    .push(graph_links.offsets.len() as u64 - 1);
-                converter.iterate_level_points(level, |_, links| {
-                    graph_links.links.extend_from_slice(links);
-                    graph_links.offsets.push(graph_links.links.len() as u64);
-                    links.clear();
-                });
-            }
-            graph_links.reindex = converter.reindex;
+    fn from_converter(converter: GraphLinksConverter) -> OperationResult<Self> {
+        if let Some(path) = converter.path {
+            GraphLinksRam::load_from_file(&path)
```

Comment:
Avoid memory duplication, if possible

Reply:
Does `load_from_file` is called when we convert from `GraphLayersBuilder::into_graph_layers`. It means that we save data to disk and load this data in RAM mode. Is it correct behavior? I tried to avoid such situation

Reply:
hm, yes, might make sense to change. Will do memory conversion